src/ORM/DataQuery.php

```diff
@@ -723,6 +721,56 @@ public function conjunctiveGroup()
         return new DataQuery_SubGroup($this, 'AND', $clause);
     }
 
+    /**
+     * Adds a Common Table Expression (CTE), aka WITH clause.
+     *
+     * Use of this method should usually be within a conditional check against DB::get_conn()->supportsCteQueries().
+     *
+     * @param string $name The name of the WITH clause, which can be referenced in any queries UNIONed to the $query
+     * and in this query directly, as though it were a table name.
+     * @param string[] $cteFields Aliases for any columns selected in $query which can be referenced in any queries
+     * UNIONed to the $query and in this query directly, as though they were columns in a real table.
+     * NOTE: If $query is a DataQuery, then cteFields must be the names of real columns on that DataQuery's data class.
+     * @param string|string[] $onClause The "ON" clause (escaped SQL statement) for an INNER JOIN on the query.
+     * It can either be a full clause (like you would pass to {@link leftJoin()} or {@link innerJoin()}),
+     * or it can be an array mapping of the field(s) on the dataclass table that map with the field(s) on the CTE table
+     * e.g. ['ID' => 'cte_id']
+     * If you want to use another join type, leave this blank and call the appropriate join method.
+     */
+    public function with(string $name, DataQuery|SQLSelect $query, array $cteFields = [], string|array $onClause = '', bool $recursive = false): static
+    {
+        $schema = DataObject::getSchema();
+
+        // If the query is a DataQuery, make sure all manipulators, joins, etc are applied
```

Comment:
This blocks confusing me, because it's setting select fields on the SQLSelect if you pass in a DataQuery.   Shouldn't the DataQuery already do that when you call DataQuery::query()?

Reply:
It's setting the select fields to _only_ be the fields passed in as `$cteFields`, overriding the explicit _every single column_ selection that `DataQuery` does by default, which would cause issues if you're declaring specific `$cteFields` since those are (presumably) the fields you're selecting in your recursive query.

If you have [some query] union [some other query] for a recursive CTE, both of those queries must have the same number of columns selected (at least in MySQL).

Reply:
Ah OK could you update the inline comment to say to summarise that it's overriding what DataQuery does by default

Reply:
Done.